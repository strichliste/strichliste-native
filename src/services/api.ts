const API_URL = "https://demo.strichliste.org/api/";

export async function fetchJson(
  endpoint: string,
  options: RequestInit = {}
): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function post(endpoint: string, params: any): Promise<any> {
  const options: RequestInit = {
    mode: "cors",
    method: "POST",
    body: JSON.stringify(params),
    headers: { "Content-Type": "application/json; charset=utf-8" }
  };
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function get(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`).then(async res => res.json());
}

export async function restDelete(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, {
    method: "delete"
  }).then(async res => res.json());
}
