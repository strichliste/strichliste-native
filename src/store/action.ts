import {
  ArticleActions,
  ErrorActions,
  LoaderActions,
  SearchActions,
  SettingsActions,
  TransactionActions,
  UserActions,
} from './reducers';

export type Action =
  | ArticleActions
  | ErrorActions
  | LoaderActions
  | SettingsActions
  | SearchActions
  | TransactionActions
  | UserActions;
