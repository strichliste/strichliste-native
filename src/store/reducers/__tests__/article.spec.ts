/* eslint-disable @typescript-eslint/no-explicit-any */

import { DeepPartial } from 'redux';
import { get, post } from '../../../services/api';
import { getMockStore } from '../../../spec-configs/mock-store';
import { Action } from '../../action';
import {
  article,
  articlesLoaded,
  getArticle,
  getArticleByBarcode,
  getArticleList,
  startAddArticle,
  startLoadingArticles,
} from '../article';

jest.mock('../../../services/api', () => ({
  get: jest.fn(),
  post: jest.fn(),
}));

describe('article reducer', () => {
  let action: DeepPartial<Action>;

  describe('with non matching action', () => {
    it('returns initial state', () => {
      const initialState = {};
      expect(article(undefined, {} as any)).toEqual(initialState);
    });
  });

  describe('with a articlesLoaded action', () => {
    const articlesResponse = [
      {
        id: 2,
        name: 'Club Mate',
        barcode: '13373243',
        amount: 100,
        active: true,
        usageCount: 0,
        precursor: null,
        created: '2018-08-17 14:21:25',
      },
    ];

    beforeEach(() => {
      action = articlesLoaded(articlesResponse as any);
    });

    it('creates a new article entry in the store', () => {
      const expectedResult = {
        2: {
          id: 2,
          name: 'Club Mate',
          barcode: '13373243',
          amount: 100,
          active: true,
          usageCount: 0,
          precursor: null,
          created: '2018-08-17 14:21:25',
        },
      };
      const result = article(undefined, action as Action);
      expect(result).toEqual(expectedResult);
    });
  });
});

describe('action creators', () => {
  describe('startAddArticle', () => {
    it('', async () => {
      (post as any).mockImplementationOnce(() =>
        Promise.resolve({ article: { id: 1 } })
      );
      const store = getMockStore();
      await startAddArticle(store.dispatch, { id: 1 } as any);
      expect(post).toHaveBeenCalledWith('article', { id: 1 });
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('getArticleByBarcode', () => {
    it('fetches articles by barcode', async () => {
      (get as any).mockImplementationOnce(() =>
        Promise.resolve({ articles: [{ id: 1 }] })
      );
      const store = getMockStore();
      await getArticleByBarcode(store.dispatch, 'asdf');
      expect(get).toHaveBeenCalledWith('article?barcode=asdf');
      expect(store.getActions()).toMatchSnapshot();
    });

    it('throws no articles error if the result list is empty', async () => {
      (get as any).mockImplementationOnce(() =>
        Promise.resolve({ articles: [] })
      );
      const store = getMockStore();
      try {
        await getArticleByBarcode(store.dispatch, 'asdf');
      } catch (error) {
        expect(error.message).toBe('no articles are matching the barcode');
      }
      expect(get).toHaveBeenCalledWith('article?barcode=asdf');
    });
  });

  describe('startLoadingArticles', async () => {
    (get as any).mockImplementationOnce(() =>
      Promise.resolve({ articles: [{ id: 1 }] })
    );

    const store = getMockStore();
    await startLoadingArticles(store.dispatch);
    expect(get).toHaveBeenCalledWith('article');
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('selectors', () => {
  describe('getArticle', () => {
    it('return the article state', () => {
      expect(
        getArticle({
          article: { 1: { id: 1 } },
        } as any)
      ).toEqual({ 1: { id: 1 } });
    });
  });
  describe('getArticleList', () => {
    it('returns all active articles as array', () => {
      expect(
        getArticleList({
          article: {
            1: { id: 1, isActive: true },
            2: { id: 2, isActive: false },
          },
        } as any)
      ).toEqual([{ id: 1, isActive: true }]);
    });
  });
});
