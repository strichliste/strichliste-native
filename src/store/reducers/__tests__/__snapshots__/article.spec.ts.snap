// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators getArticleByBarcode fetches articles by barcode 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "articlesLoaded",
  },
]
`;

exports[`action creators startAddArticle  1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 1,
      },
    ],
    "type": "articlesLoaded",
  },
]
`;
